samza-kv/src/test/java/org/apache/samza/storage/kv/TestLargeMessageSafeStore.java

```diff
@@ -0,0 +1,180 @@
+    LargeMessageSafeStore largeMessageSafeKeyValueStore = new LargeMessageSafeStore(store, storeName, false, maxMessageSize);
+    byte[] key = new byte[16];
+    byte[] smallMessage = new byte[32];
+
+    List<Entry<byte[], byte[]>> entries = new ArrayList<>();
+    entries.add(new Entry<>(key, smallMessage));
+
+    largeMessageSafeKeyValueStore.putAll(entries);
+    Mockito.verify(store).putAll(Matchers.eq(entries));
+    Mockito.verify(store, Mockito.never()).put(Matchers.any(byte[].class), Matchers.any(byte[].class));
+  }
+
+  @Test
+  public void testLargeMessagePutAllFailureWithDropLargeMessageDisabled() {
+    LargeMessageSafeStore largeMessageSafeKeyValueStore = new LargeMessageSafeStore(store, storeName, false, maxMessageSize);
+    byte[] key = new byte[16];
+    byte[] largeMessage = new byte[maxMessageSize + 1];
+
+    List<Entry<byte[], byte[]>> entries = new ArrayList<>();
+    entries.add(new Entry<>(key, largeMessage));
+
+    try {
+      largeMessageSafeKeyValueStore.putAll(entries);
+      Assert.fail("The test case should have failed due to a large message being passed to the changelog, but it didn't.");
+    } catch (SamzaException e) {
+      Mockito.verifyZeroInteractions(store);
+    }
+  }
+
+  @Test
+  public void testSmallMessagePutWithSerdeAndDropLargeMessageDisabled() {
+    LargeMessageSafeStore largeMessageSafeKeyValueStore = new LargeMessageSafeStore(store, storeName, false, maxMessageSize);
+
+    Serde<Long> longSerde = new LongSerde();
+    long longObj = 1000L;
+    byte[] key = longSerde.toBytes(longObj);
+
+    JsonSerdeV2<Map<String, Object>> jsonSerde = new JsonSerdeV2<>();
+    Map<String, Object> obj = new HashMap<>();
+    obj.put("jack", "jill");
+    obj.put("john", 2);
+    byte[] smallMessage = jsonSerde.toBytes(obj);
+
+    List<Entry<byte[], byte[]>> entries = new ArrayList<>();
+    entries.add(new Entry<>(key, smallMessage));
+
+    largeMessageSafeKeyValueStore.putAll(entries);
+    Mockito.verify(store).putAll(Matchers.eq(entries));
+    Mockito.verify(store, Mockito.never()).put(Matchers.any(byte[].class), Matchers.any(byte[].class));
+  }
+
+  @Test
+  public void testLargeMessagePutWithSerdeAndDropLargeMessageDisabled() {
+    LargeMessageSafeStore largeMessageSafeKeyValueStore = new LargeMessageSafeStore(store, storeName, false, maxMessageSize);
+
+    Serde<Long> longSerde = new LongSerde();
+    long longObj = 1000L;
+    byte[] key = longSerde.toBytes(longObj);
+
+    Serde<String> stringSerde = new StringSerde();
+    String largeString = StringUtils.repeat("a", maxMessageSize + 1);
+    byte[] largeMessage = stringSerde.toBytes(largeString);
+
+    List<Entry<byte[], byte[]>> entries = new ArrayList<>();
+    entries.add(new Entry<>(key, largeMessage));
+
+    try {
+      largeMessageSafeKeyValueStore.putAll(entries);
+    } catch (SamzaException e) {
+      Mockito.verifyZeroInteractions(store);
+    }
+  }
+
+  @Test
+  public void testPutSuccessWithDropLargeMessageEnabled() {
```

Comment:
Consider also adding a test case where the message is small and drop large message is enabled, making sure that the message is put into the store.

Reply:
Done.